File: Python/leetcode_20.py
'''
Given a string s containing just the characters
'(', ')', '{', '}', '[' and ']', determine if the input string is valid.

An input string is valid if:

Open brackets must be closed by the same type of brackets.
Open brackets must be closed in the correct order.
Every close bracket has a corresponding open bracket of the same type.
'''


def is_alid(s: str) -> bool:
    stack = []
    bracket_map = {')': '(', '}': '{', ']': '['}

    for char in s:
        if char in bracket_map.values():  # Opening brackets
            stack.append(char)
        elif char in bracket_map:  # Closing brackets
            if not stack or stack[-1] != bracket_map[char]:
                return False
            stack.pop()
        else:
            return False  # Invalid character

    return not stack

File: JavaScript/leetcode_20.js
var isValid = function (s) {
    let stack = [];
    const bracket_map = { ")": "(", "}": "{", "]": "[" };

    for (let char of s) {
        if (Object.values(bracket_map).includes(char)) {
            stack.push(char);
        } else if (char in bracket_map) {
            if (
                stack.length === 0 ||
                stack[stack.length - 1] !== bracket_map[char]
            ) {
                return false;
            }
            stack.pop();
        }
    }

    return stack.length === 0;
};

console.log(isValid("()"));
